### ISSUE TITLE
Rule Proposal: `key-linebreak` enforce linebreaks after object key

### ISSUE THREAD

**Issue by sharmilajesupaul** [eslint/eslint#9699]
### `key-linebreak`

**Please describe what the rule should do:**
This rule enforces consistency in linebreaks after an object key.

For context: we've come across places in our code where linebreaks occur after object keys, usually as a result of using an autoformatting tool or to avoid violating [`max-len`](https://eslint.org/docs/rules/max-len). 

We would like to make a decision on allowing this code style and it would be great if we could enforce consistent behavior using eslint.

**Options:** 
This rule has a string option:
- `"beside"`: (default) requires that an object key has a continuation of the value on the same line
- `"below"`: requires that an object key is followed by a linebreak before the value

**What category of rule is this? (place an "X" next to just one item)**

- [X] Enforces code style
- [ ] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

`"beside"` option (default):
```jsx
// 👍  good
this.state = {
  scanMethod: flowContextLongLongName === FLOW_CONTEXTS.SUSPENSION_APPEAL
    ? DEVICE_OPTIONS.CURRENT_DEVICE
    : DEVICE_OPTIONS.MOBILE_APP,
  scanMethod1: 'string ',
  [dynamic]: 'foo',
  parens: (
    something === somethingElse
  ),
}

// 👎  bad
this.state = {
  scanMethod:
    flowContextLongLongName === FLOW_CONTEXTS.SUSPENSION_APPEAL
      ? DEVICE_OPTIONS.CURRENT_DEVICE
      : DEVICE_OPTIONS.MOBILE_APP,
  scanMethod1:
    'string ',
  [dynamic]:
    'foo',
  parens:
    (
      something === somethingElse
    ),
}
```
`"below"` option:
```jsx
// 👍  good
this.state = {
  scanMethod:
    flowContextLongLongName === FLOW_CONTEXTS.SUSPENSION_APPEAL
      ? DEVICE_OPTIONS.CURRENT_DEVICE
      : DEVICE_OPTIONS.MOBILE_APP,
  scanMethod1:
    'string ',
  [dynamic]:
    'foo',
  parens:
    (
      something === somethingElse
    ),
}

// 👎  bad
this.state = {
  scanMethod: flowContextLongLongName === FLOW_CONTEXTS.SUSPENSION_APPEAL
    ? DEVICE_OPTIONS.CURRENT_DEVICE
    : DEVICE_OPTIONS.MOBILE_APP,
  scanMethod1: 'string ',
  [dynamic]: 'foo',
  parens: (
    something === somethingElse
  ),
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This rule will affect code style in a very broad way and it doesn't seem like there are any plugins that will be suitable for this case.

cc. @ljharb

**Comment by ljharb** [comment]
Another alternative is adding an option to `key-spacing` that includes linebreaks in the "spacing"

**Comment by ilyavolodin** [comment]
Would this only apply to object definitions? What about destructuring with renaming? Also, I could see that if this rule is implemented we would get a bunch of requests for exceptions for things like  inline objects (`callFn({foo: bar});`) and exceptions based on number of properties/length.
I think I would prefer it as an option under `key-spacing`. If it would be a separate rule, I would probably also suggest adding `consistent` option (as in, don't care if it's beside or below, as long as it's the same within the same object definition).

**Comment by ljharb** [comment]
Since it deals with keys, it could deal with anything that has them - including destructuring patterns. However, I'd expect "object literals" and "destructuring patterns" to be governed by separate options.

As for renaming, that's not a key, so I'm not sure that needs to apply.

An option under `key-spacing` sounds fine - and "consistent" also sounds great. What more do we need for a PR that does that to be accepted?

**Comment by sharmilajesupaul** [comment]
Adding this to `key-spacing` as an option sounds good to me too. And I think having a `consistent` option would be useful in that case as well. 

@ljharb, @ilyavolodin, you bring up a good point regarding inline objects. The `"below"` option would not allow a developer to write any inline objects and that wasn't our intention. Would it make sense to add an option to ignore inline objects when using `"below"`?

**Comment by ljharb** [comment]
Perhaps instead, we could make this new option entirely ignore single-line objects, since it only really makes sense for multiline objects?

**Comment by ljharb** [comment]
If so, then an option name like `multiLineObjectValuePlacement` or something might work, with possible values of "below", "beside", "consistent", and "ignore" (and "ignore" would be the default, for back compat)?

**Comment by not-an-aardvark** [comment]
It seems like the `key-spacing` rule already has a `multiLine` option for specifying options which are specific to multi-line objects. Would it work to add this configuration option to there, rather than adding a top-level configuration that ignores single-line objects?

**Comment by sharmilajesupaul** [comment]
@not-an-aardvark thanks for suggesting this, that should work well for this use case.

We could add configuration to `key-spacing`'s `multiline` option called `linebreak` that takes "below", "beside", and "consistent" as values?

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.

**Comment by ljharb** [comment]
@nzakas in this case, it's not a lack of interest, it's that the eslint team didn't respond. @not-an-aardvark, any chance you could respond to https://github.com/eslint/eslint/issues/9699#issuecomment-360927469 so we can reopen this?

**Comment by not-an-aardvark** [comment]
Sorry, I had interpreted https://github.com/eslint/eslint/issues/9699#issuecomment-360927469 as a modification to the proposal, not a direct question. I had also left a :+1: on the comment to indicate my assent.

To me, it doesn't seem like the open question would have added so much ambiguity that it would have prevented the team from evaluating the core of the proposal, so my inclination is to leave this closed. However, I'd be fine with reopening it if anyone on the team feels differently.

**Comment by ljharb** [comment]
This option is needed to help allow eslint to cover a formatting style that prettier is otherwise required for. It would be great if it could be reopened.

**Comment by nzakas** [comment]
I wouldn't reopen unless there is a commitment from someone to do an implementation. Right now, we just don't have the team bandwidth to deal with this otherwise.

**Comment by ljharb** [comment]
@nzakas we’ll be happy to work on the implementation once the proposal is accepted.

**Comment by nzakas** [comment]
Reopening per last comment.

**Comment by ljharb** [comment]
We'll still be happy to work on the implementation once the proposal is accepted, and this is still a concern for the Airbnb styleguide. Please reopen it?
